### ISSUE TITLE
Implement Responsive User Location Tracking Mode

### ISSUE THREAD

**Issue by bleege** [mapbox/mapbox-gl-native#2049]
The User Location tracking mode (None, Follow, Bearing) was disabled in #1856 as the User Location dot was being updated on the main UI thread.  The core logic is there, but it needs to be implemented in a more efficient way so that the UI remains responsive.

/cc @erf

**Comment by erf** [comment]
Could it be added as a annotation?

**Comment by ljbade** [comment]
@bleege Is the issue that there is too many events?

**Comment by erf** [comment]
i made a lowpass filter in  #1968 which could be used

**Comment by erf** [comment]
in addition to some minimal time / angle diff

**Comment by erf** [comment]
I think you should move out the setDirection call from updateMap to the onCompassChange event.

**Comment by bleege** [comment]
> @bleege Is the issue that there is too many events?

Yes.

**Comment by ljbade** [comment]
@bleege How does iOS handle that?

**Comment by 1ec5** [comment]
@ljbade, on iOS, MGLMapView doesn’t do anything special. But it seems that Core Location is doing a better job of coalescing location updates than what we’re using on Android: #1676.

**Comment by bleege** [comment]
> But it seems that Core Location is doing a better job of coalescing location updates than what we’re using on Android

Exactly.  We need to put a regulator / distance filter on updates to location or specifically that a certain amount of distance / time needs to elapse before updating the dot on the map.

**Comment by ljbade** [comment]
@bleege I recall the LOST framework has this ability to filter (certainly Google Play services do).

We still need to work in @erf's compass filter for that data.

**Comment by ljbade** [comment]
I was getting confused between location/GPS events and map position change events.

**Comment by ljbade** [comment]
In #2409 I'm cleaning up the `updateMap()` code so this should be unblocked by performance issues now.

**Comment by mikemorris** [comment]
Have you tried sampling input/direction (at a constant rate of 15 times per second or so) instead of trying to process or filter all events? I've done this in the past for handling user input from a gyroscope and accelerometer in JavaScript, where it was previously choking on large bursts of minor change events.

Sample code at https://github.com/mikemorris/gyrocopter/blob/master/gyrocopter.js

**Comment by ljbade** [comment]
Hmm I might have accidentally removed the user tracking stuff in #2002, but the code is still there in git history so perhaps we can have a go at making it really good this time.

Lets start with just location/GPS (and heading?) tracking and work on compass after.

Should we also hit the UI side of getting our GPS dot looking better?

**Comment by ljbade** [comment]
Im thinking we eventually want a few modes (start with first two):
- no tracking
- just location tracking
- location tracking and map heading/rotation from GPS direction
- location tracking and map heading/rotation from rate limited compass

Also do we make touch events cancel tracking (separate for rotation and pan events). Or leave that up to the app via the listening events.

**Comment by tobrun** [comment]
Discussed with @bleege, 
Important part here is that while zooming the GPS marker 'jitters' and shown GPS location is wrong. 
If that problem is identified and fixed, I will start looking at optimising the bearing part.

**Comment by tobrun** [comment]
While zoomed in I'm above the city Leuven and the orange highway

![device-2015-10-08-113858](https://cloud.githubusercontent.com/assets/2151639/10362861/4e6259fe-6db1-11e5-9b45-a148a8a0faaa.png)

Zooming out, I'm below the city Leuven and highway:

![device-2015-10-08-113821](https://cloud.githubusercontent.com/assets/2151639/10362862/4e664c1c-6db1-11e5-8a02-1adebdb7dc04.png)

**Comment by tobrun** [comment]
I guessing of the reasons why its jitters can be found in setLayoutParams and calling request layout:
```java
    // Update Location
            FrameLayout.LayoutParams lp = new FrameLayout.LayoutParams((int) iconSize, (int) iconSize);
            lp.leftMargin = (int) (screenLocation.x - iconSize / 2.0f);
            lp.topMargin = (int) (screenLocation.y + iconSize / 2.0f);
            mGpsMarker.setLayoutParams(lp);
            rotateImageView(mGpsMarker, 0.0f);
            mGpsMarker.requestLayout();
```


This can be replace by setX and setY:
```
            mGpsMarker.setX(screenLocation.x);
            mGpsMarker.setY(screenLocation.y);
```

**Comment by tobrun** [comment]
I'm guessing that wrong location is related to not taking into account the size of GPS marker itself.

**Comment by tobrun** [comment]
I'm glad to report that location bug is resolved! 
The marker location did not take in account the size of the marker.

![device-2015-10-08-122358](https://cloud.githubusercontent.com/assets/2151639/10363903/b2fa72b0-6db7-11e5-9f66-94fd177b3413.png)

![device-2015-10-08-122417](https://cloud.githubusercontent.com/assets/2151639/10363907/b6478db8-6db7-11e5-9106-6b30386e938d.png)

**Comment by tobrun** [comment]
Things I did to improve performance:
 - avoid creating objects in update method (this is called multiple times while doing gestures)
 - move some code to initialisation (some things were set every time, while once would be enough)

**Comment by tobrun** [comment]
Ah, now I'm noticing the real problem related to this issue, 
when zoomed in the user location dot jumps from location to location.

**Comment by bleege** [comment]
Mapbox for iOS has this pretty well figured out.  We should start this improvement by porting the logic that they used.  Details can be found in the links below.  For those new to iOS, `MGLUserLocation.m` and `MGLUserLocationAnnotationView.m` are going to be the ones to start looking at.

* [MGLUserLocation.m](https://github.com/mapbox/mapbox-gl-native/blob/8b6f928718961306a2b7bfdfc5f8decf0d0ddf7a/platform/ios/MGLUserLocation.m)
* [MGLUserLocationAnnotationView.h](https://github.com/mapbox/mapbox-gl-native/blob/8b6f928718961306a2b7bfdfc5f8decf0d0ddf7a/platform/ios/MGLUserLocationAnnotationView.h)
* [MGLUserLocationAnnotationView.m](https://github.com/mapbox/mapbox-gl-native/blob/8b6f928718961306a2b7bfdfc5f8decf0d0ddf7a/platform/ios/MGLUserLocationAnnotationView.m)
* [MGLUserLocation_Private.h](https://github.com/mapbox/mapbox-gl-native/blob/8b6f928718961306a2b7bfdfc5f8decf0d0ddf7a/platform/ios/MGLUserLocation_Private.h)

**Comment by tobrun** [comment]
We are currently using the same thresholds as iOS (1m) but the data we are receiving is just not as accurate. Will do some testing tomorrow by fiddling with LOST configurations.
If that doesn't workout I will try to do some different strategies tomorrow.

**Comment by bleege** [comment]
Ok, I tracked down [the code in iOS for handling this](https://github.com/mapbox/mapbox-gl-native/blob/4f6f2c19fd58f07a39a43c8547fe75534ec12905/platform/ios/MGLMapView.mm#L2686-L2765).  In iOS `didUpdateLocations` method is the moral equivalent of `onLocationChanged()` in Android.  Conceptually we should be able to port the same logic and avoid reinventing the wheel.

**Comment by bleege** [comment]
Oh nevermind me.  That's the code for getting the UserDot to not be so jumpy.

**Comment by ljbade** [comment]
@tobrun - ah I think what happened here was that at some stage we changed the GPS dot image but forgot to update the code! Nice catch.

**Comment by tobrun** [comment]
I will first take a look reintegrating 
> location tracking and map heading/rotation from GPS direction
location tracking and map heading/rotation from rate limited compass

before optimising the location updates. 
I feel like the optimisation should be streamlined in LOST itself instead fixing it ourselves.

**Comment by tobrun** [comment]
I'm going to replace following enum
```java
public enum UserLocationTrackingMode {
        NONE, FOLLOW, FOLLOW_BEARING
}
```
with integers while using an IntRange annotation

**Comment by tobrun** [comment]
I added location change animation

**Comment by tobrun** [comment]
If we enable location tracking, we should disable updating the Location marker hence it will always be centered in those cases. (this will remove jitter while zooming)

**Comment by tobrun** [comment]
Calling next method for animating location change conflicts with output from gestures
```java
setCenterCoordinate(location, true /**animated**/);
```
Note: this is a different animation than mentioned above for non locational tracking

**Comment by tobrun** [comment]
Done:
 - reintegrated tracking modes
 - added animation for user dot for non tracking mode
 - changed the way we handle different types of tracking modes. 
  - while tracking I'm not updating the userdot location (stays in center screen) 
  - provide a way to handle both a dirty as non dirty system (animations)
 - adapted enabling/disabling gestures for different types of tracking modes
 - fixed bug showing wrong location
 - implemented alternative for enums

TODO:
 - read correct compass data from sensors (current impl. gets data from tilting device)
 - improve location updates
 - enable animation for bearing tracking mode for location update
 - enable animation for bearing tracking mode for bearing update
 - replace bearing GPS marker icon
 - cleanup code

Known bugs:
- tracking mode using compass does not work (see first TODO)
- on configuration change the userdot is not correctly positioned while using a tracking mode
  - MapView is not measured yet while using the width/height to postion GpsMarker

@ljbade : need to ask you some things related to enable animation, will ping Monday afternoon.

cc @bleege

**Comment by bleege** [comment]
@tobrun I really like the Enum alternative that you found.  The key for me was that it still allowed for the ease of use (see below).  It may be a bit kludgy to setup but that's on our side and not on developers who use the Mapbox SDK.  Still makes you wonder why Google didn't provide better compile time support for Enums instead of introducing an annotation.

```java
mMapView.setUserLocationTrackingMode(MapView.TRACKING_NONE);
```

**Comment by bleege** [comment]
While testing the latest from [`2049-Responsive-User-Tracking`](https://github.com/mapbox/mapbox-gl-native/compare/2049-Responsive-User-Tracking) I'm getting an incorrect bearing when using either `MapView.TRACKING_FOLLOW_BEARING_GPS` or `MapView.TRACKING_FOLLOW_BEARING_COMPASS` and then rotating the device 360 degrees.  In other words arrow doesn't move in the direction I'm facing (see first screenshot).  Secondly, when I rotate the app the user dot gets stuck in the upper left hand corner of the Map (see second screenshot).

![device-2015-10-09-141403](https://cloud.githubusercontent.com/assets/989928/10403270/0b967bca-6e90-11e5-8a4d-365e0116cd9e.png)
_Consistently Points One Way_

![device-2015-10-09-142158](https://cloud.githubusercontent.com/assets/989928/10403461/611b23d8-6e91-11e5-92da-52703519d81e.png)
_Stuck After Rotation_

**Comment by tobrun** [comment]
> not switching MapView.TRACKING_FOLLOW_BEARING_GPS   when rotating device

this is correct, you will only be updated when location has changed, 
rotating device will probably not trigger this.

> MapView.TRACKING_FOLLOW_BEARING_COMPASS

as indicated this is not yet implemented.

**Comment by ljbade** [comment]
@bleege @tobrun Interesting about enums. I had always wondered why the Android framework exclusively used static ints instead of enums.

@tobrun Perhaps you should only switch from the circle dot to the arrow after receiving the first location update.

**Comment by tobrun** [comment]
If location tracking with bearing is enabled.
 - Should the map change direction in the way you are heading? 
 - only the GPS marker?

For the former I need to discuss some things with @ljbade (how to hook into surface view animations while handling gestures?) This is also important for getting velocity/scroller/interpolator things in gestures.

cc @bleege

**Comment by tobrun** [comment]
Need to optimise data fed to animation
https://youtu.be/1lJec_vfHjs

**Comment by tobrun** [comment]
In motion sensor docs they advice using a [Rotation vector](http://developer.android.com/reference/android/hardware/Sensor.html#TYPE_ROTATION_VECTOR
) sensor over current setup:
> For example, the rotational vector sensor is ideal if you are developing a game, an augmented reality application, a **2-dimensional** or 3-dimensional **compass**, or a camera stabilisation app. 

> In most cases, using these sensors is a **better choice than using the accelerometer and geomagnetic field sensor or the orientation sensor**.

**Comment by ljbade** [comment]
@tobrun Hmm does that precalculate the accelerometer + compass stuff for us? Might be worth testing.

**Comment by ljbade** [comment]
@tobrun Hmm does that precalculate the accelerometer + compass stuff for us? Might be worth testing.

**Comment by bleege** [comment]
> If location tracking with bearing is enabled.
* Should the map change direction in the way you are heading?
* only the GPS marker?

@tobrun Yes, the map should change and the GPS marker should indicate that "bearing mode" is on.

**Comment by tobrun** [comment]
> Hmm does that precalculate the accelerometer + compass stuff for us? Might be worth testing.

We still need to calculate. Difference is that it uses a gyroscope to produce more stable azimuth values. 

After milestone 2.1.0:
We need a game plan to handle different algorithms/devices. Not all Android devices will have the sensors mentioned above (see [Compatibility definition Android](https://static.googleusercontent.com/media/source.android.com/en//compatibility/android-cdd.pdf)).

**Comment by ljbade** [comment]
@tobrun Ah yeah thats true. Might need fallback.. Be nice if all this code was part of LOST: https://github.com/mapzen/LOST/issues/30

**Comment by tobrun** [comment]
I added a extra user tracking mode that combines the compass and GPS bearing. As @ljbade indicated we should expose them separately but users will also expect to have a fine tuned combined version.

**Comment by ljbade** [comment]
@tobrun Perfect, gives developers complete control over how they want there app to behave.

**Comment by tobrun** [comment]
I moved some things out of this branch needed for android 2.1.0. 
Currently moved this to 2.2.0 milestone. 

What is required to finish this ticket:
  - smooth animation when calling setBearing(degrees, true /* animated */)
    - @incanus indicated that we need to sync. frames. @ljbade could you help me out with this?
  - correct image + rotation for bearing

**Comment by ljbade** [comment]
@tobrun Excellent work!

Yeah I can help you with frame sync, I digged into this a while ago but still havn't come a good solution yet. Part of the problem is that TextureView seems to add a one frame lag, and various methods to sync with that that I have tried so far did not work.

**Comment by ljbade** [comment]
@tobrun I have been playing with Android systrace to get an idea of what happens in our app.

Below is a screen shot with annotations showing our current process to draw a frame to the screen. To record this I had the map open at location with GPS marker, I then rotated the map, started recording, and tapped the compass to start the reset to north animation. Thus there is no gestures involved in this capture.

![render](https://cloud.githubusercontent.com/assets/1215089/10567901/07f7c9f6-765b-11e5-90d5-4192d3607c75.png)

**Comment by ljbade** [comment]
Comparing iOS to Android's flow, iOS updates the user location after `renderSync`, but on Android we are calling it before. Going to switch this around to see what happens.

**Comment by ljbade** [comment]
Switching the order to match iOS did not seem to make much difference.

**Comment by ljbade** [comment]
I did an experiment with the Android screen video recorder. I'll post a Dropbox link to the video once it finishes uploading.

I set the app to rotate the map by 90 degrees every frame with the GPS marker on.

What I observe is that the compass and user location marker are synced correctly with the map after switching the order to match iOS (see above comments).

However we are only outputting every second rendered map frame to the screen. So the map runs at 30fps, whereas the user location runs at 60fps. This is what is making the marker jump.

**Comment by ljbade** [comment]
Been trying a few things. First I made the `new Runnable` only happen once per app to reduce memory churn.

I tried removing the `mDirty` check and this causes core GL to start pumping out at 120fps, but holds up UI thread so much not many of these frames ever make it to the screen.

**Comment by ljbade** [comment]
Decided to try just calling renderSync at 60 times per second, but it still only renders every second frame.

**Comment by ljbade** [comment]
Aha I think I'm on to something here. If I run renderSync on a loop on another thread then post an updateCompass/updateGpsMarker function to the UI thread I get smooth panning and properly synced GPS marker.

**Comment by ljbade** [comment]
So far I have got `renderSync` on a new thread. After each render the thread then posts on the UI thread.

Initially the render thread was uncapped and rendering at 200fps with no lag between the map and the marker. However this heats the phone up and reduces battery life.

I then implemented a 60fps cap using Choreographer. This reduces CPU use, but now there is the occasional jump in the GPS marker.

I tried synchronising with the UI thread to wait for marker update to control the exact timing but this reduces the frame rate to 30fps even with just a `synchronized` block.

We are going need a better method for updating the marker. I am thinking we need to capture the map's transform etc (or calculate the marker position and compass rotation) in the rendering thread, then pass that data to the UI thread.

**Comment by ljbade** [comment]
@tobrun I have an idea for how to smoothly animate marker updates. The trick will be to animate the lat/lng coords of the marker, not the screen position. Then when we update the marker position each frame with the map transform we will be animating the marker and responding to gestures at the same time.

**Comment by tobrun** [comment]
>  I have an idea for how to smoothly animate marker updates. The trick will be to animate the lat/lng coords of the marker, not the screen position. Then when we update the marker position each frame with the map transform we will be animating the marker and responding to gestures at the same time.

 :+1:
Great finds and excited to get this in! 
I'm curious about how this will look in code.

**Comment by ljbade** [comment]
@tobrun First I think we need to create a separate GPSMarkerView. Then we can implement various methods in it for using the Android animation interpolation classes to update the lat/lng it holds. Then just use the lat/lng in toScreenPosition as part of updateGpsMarker.

**Comment by ljbade** [comment]
Improvements based on my findings above is in #2670 

Been thinking more about getting the marker sync.

Basic Android code flow (from `Choreographer` class source code):
1. vsync event fires (hardware based on screen refresh)
2. Android collects, processes and delivers all touch events - we call various functions on NativeMapView to update the map's "transform" - ie the current view
3. Android processes all Android animation framework updates (things such as the `View.animate()` methods - here is where possibly the GPS marker animation would update
4. Android processes all custom `Choreographer` callbacks - this is where our `RenderThread` calls `NativeMapView.renderSync()` which is where the actual GL code renders to the `SurfaceTexture` and sends it to `TextureView` via `eglSwapBuffers()` - this is rendered with the current "transform" updated in step 2
5. Android performs `View` layout hierarchy traversal, calculates updated positions of `View`s and their properties
6. Android calls each `View`s `onDraw()` method to render their updated contents - this is where our `TextureView` fetches the latest GL rendered `SurfaceTexture` and blits it to the screen

@bleege @tobrun Let me know if you can follow ^

Now the trick here is that we don't know the final map "transform" until step 4 at `renderSync`. If I follow the iOS code, this final calculation happens in `renderSync`.

However if we don't update the GPS marker position before step 5 then we will miss the deadline and the marker position will get updated in the *next frame* - which will break the sync.

The complication here is that we can only update the Android UI on the UI thread, not on the thread that calls `renderSync`.

**Comment by ljbade** [comment]
Here is the new systrace from #2670 
![screenshot from 2015-10-19 21 54 36](https://cloud.githubusercontent.com/assets/1215089/10576747/6e8d89d4-76b1-11e5-85b7-17f715864a5a.png)

This contradicts my thoughts from the last comment :-/

First, notice how long the core GL rendering takes, this would explain why we were getting 30fps before as waiting for core GL on the UI thread was pushing us over one frame.

Now look at when we update the GPS marker. This results in 2 frame lag between when we receive the input events and finally get the updated GPS marker on the screen.

The map itself will only have 1 frame of lag (this would be 0 if we finished rendering fast enough).

Thus we have a mismatch.

We need to get the GPS marker update down to 1 frame lag.

**Comment by ljbade** [comment]
Another issue is when using a gesture there are quite a few frame hitches. It appears to be something to do with "queueBuffer" and "dequeueBuffer" which I think is from SurfaceTexture.

**Comment by ljbade** [comment]
Found an even simpler solution to 60fps.

I enable `onDraw` in `MapView` via `setWillNotDraw(false)` (since by default `ViewGroup` classes do not draw).

Then I run `renderSync` and `updateGpsMarker` inside the `onDraw`. As `MapView` is higher in the hierachy then the `TextureView`, compass, and GPS marker it will get called first.

There is still a 1 frame lag between the map and GPS marker however. Possibly because it is too late to update children by `onDraw`?

**Comment by bleege** [comment]
Documenting that @incanus is experimenting with this over https://github.com/mapbox/mapbox-gl-native/pull/2307

**Comment by ljbade** [comment]
For some reason ^ is not reliable, still getting 30fps.

I found https://wiki.mozilla.org/Platform/GFX/Gralloc#How_Android_is_using_Gralloc useful. It seems the problem is by default TextureView creates an async buffer queue thus we don't really know when it updates the displayed frame, and which one it used.

We do get `onSurfaceTextureUpdated()` which is called inside TextureViews draw code after it updates the frame. But how we determine which frame that was, I am not sure.

I tried to create single buffer/synchronous SurfaceTexture using some newer APIs, but was stuck with where to call `releaseTexImage()`. It seems this needs to be called from TextureView's EGL context which we have no access to.

I am beginning to think the only way we can correctly sync everything is via GL, or going back to SurfaceView and having to accept that it doesn't work with all types of Material design layouts.

**Comment by ljbade** [comment]
Good news, updating the UI in `onSurfaceTextureUpdated` has got the marker syncing, including when it skips frames (we now also skip a UI frame to match).

Now I need to work on getting smooth 60fps.

**Comment by ljbade** [comment]
Trying to combine `onSurfaceTextureUpdated` with a seperate render thread causes the sync to drop out occasionally.

Going to have to renderSync on the UI thread, but figure out why it skips every second frame.

**Comment by ljbade** [comment]
@tobrun @bleege Made a lot of progress today in `2668-sync-marker` branch.

First I refactored to create a new `UserLocationView`.

Then I changed the user location dot to be hand drawn adding a direction arrow and accuracy circle.

Here is a screenshot:
![device-2015-10-21-145441](https://cloud.githubusercontent.com/assets/1215089/10627346/ff49456e-7803-11e5-883b-9c999b0814bf.png)

**Comment by ljbade** [comment]
Next up will be animating the marker:
- location
- direction
- accuarcy
- fade in/out of the different components (and the marker itself?)

**Comment by ljbade** [comment]
Animations are done in the branch.

**Comment by tobrun** [comment]
I'm working on this again, because I have noticed 2 bugs on current progress that need to fixed asap

**Comment by tobrun** [comment]
To clarify above: 
 -  Spinning UserLocationView #3240 
 -  Stop listening to bearing #3262

**Comment by tobrun** [comment]
The new camera API from @bleege will help resolve remaining issue on this ticket: 

https://github.com/mapbox/mapbox-gl-native/pull/3396 contains the fix for animated location updates with user tracking mode enabled.

Todo list to close this issue:
 - [ ] do similar fix for `setBearing` as done for `setCenterCoordinate`
 - [ ] improve performance/jitter `UserLocationView`

Will update this ticket with related issue numbers.

**Comment by Buckley404** [comment]
Is this working correctly now? 

Currently when I try to set up tracking in the onCreate it draws the blue marker in the top left corner, digging through the code on git I found that this was because the position of the marker is being set relative to the mapview by dividing width and height by 2 (height and width would be 0 while onCreate is being called).

I altered my code to include a button I could press to start tracking in order to bypass the problem, however even though the screen is centred on the correct position, the blue dot does not draw to the screen.

In the UserLocationView class and in the update method, why are you setting the blue marker position by dividing the screen/height width by 2 to find the centre point of the screen. Why not use the latitude and longitude to draw the marker like you would with a regular addMarker call or a screen centre? 

If the code even worked properly, wouldn't moving around the map with user input reposition the blue dot in the wrong place i.e. keep it in the centre of the screen no matter where the user navigates to on the map?

**Comment by tobrun** [comment]
@Buckley404, reading through your comment I believe you are using the incorrect method. If you want the user location to show on the map in similar fashion as a marker you should call 

```java
nMapView.setMyLocationEnabled(true);
```

instead of:

```java
nMapView.setMyLocationTrackingMode(MyLocationTrackingMode.TRACKING_FOLLOW);
```

**Comment by tobrun** [comment]
#3397 landed with f80673b15c9961f089c7f341a577517f2d2aaa6c

**Comment by tobrun** [comment]
One thing I need to add to this is that current `compass` implementation is not on par with the one from Google. Current implemented code using accelerometer and magnetometer is generally spoken the accepted way of implementing a compass (as seen in books, SO and tutorials). It seems Google is adding some extra elements to the equitation or using another configuration to calculate values. One of the most notable differences is that Google's implementation is not impacted when switching perspective or tilting the device.

**Comment by Buckley404** [comment]
Thanks again @tobrun, I'll recheck my code after work however I think I was using both methods. I must have just confused myself. FYI google might just be setting minimum thresholds for when the compass should be affected by gyro.

**Comment by tobrun** [comment]
@Buckley404 I have been trying to mimic their behaviour for quite some time and I now believe that Google is doing something different than just using the sensors as we are doing.
FYI I added thresholds in latest code, we could further optimise this by implementing a low pas filter.
